lib/net.c

```diff
@@ -1038,6 +1043,29 @@ gatekeeper_get_rss_config(uint16_t port_id,
 	return ret;
 }
 
+int
+gatekeeper_setup_user(struct net_config *net_conf, const char *user)
+{
+	struct passwd *pw;
+
+	if (user == NULL) {
+		net_conf->pw_uid = getuid();
+		net_conf->pw_gid = getgid();
+		return 0;
+	}
+
+	if ((pw = getpwnam(user)) == NULL) {
+		G_LOG(ERR, "%s: failed to call getpwnam() for user %s!\n",
+			__func__, user);
+		return -1;
+	}
+
+	net_conf->pw_uid = pw->pw_uid;
+	net_conf->pw_gid = pw->pw_gid;
+
+	return initgroups(user, pw->pw_gid);
```

Comment:
`initgroups()` must be in `drop_privileges()`. Use `getlogin_r()` to find the name of the user from `net_conf->pw_uid` in `drop_privileges()`.

Once in `drop_privileges()`, we must log when `initgroups()` fails. Don't forget to add `strerror(errno)` to the log message.